### PR TITLE
fix: re-creation of existing `boundary_host_set_plugin` resource

### PR DESCRIPTION
**Pull request by elimt**
TF re-creates boundary_host_set_plugin resource even though there have been no code changes when a filter attribute has been set.

When the filter attribute is set with TF it looks like this: `{"filters":"tag:host-set=A_VALUE_AND_foo2"}` But when Boundary TF Provider reads the latest value from Boundary SDK, the value we get back looks like this: `{"filters":["tag:host-set=A_VALUE_AND_foo2"]}`. This is causing the diff checker to detect differences even though the filter value on Boundary has not changed

### CODE REVIEW

**Review by A-440Hz — APPROVED** [review]

**Review by louisruch — COMMENTED** [review]
The code her seems sane, the only thing I would love to see if we could get an updated test to validate the behavior. 
That being said I am not clear why `filters` are a special case here, perhaps @talanknight would be a better reviewer? Otherwise happy to dig into the code and discuss this path

File: internal/provider/resource_host_set_plugin.go [hunk 1]
```diff
@@ -131,6 +132,25 @@ func setFromHostSetPluginResponseMap(d *schema.ResourceData, raw map[string]inte
...
+							var flattenedFilters []string
+							for _, f := range filters {
+								if filter, ok := f.(string); ok {
+									flattenedFilters = append(flattenedFilters, filter)
+								}
+							}
+							attrMap[attrKey] = strings.Join(flattenedFilters, ",")
+						}
+					}
+				}
```
[the remark this answers is not in the document]
Comment (elimt) [reply internal/provider/resource_host_set_plugin.go:150]: I decided to iterate over it in case there are other attributes/fields in the future that will need to be changed to match what Boundary TF provider expects. There aren't a lot of attributes so I did not think iterating over it will be a big issue. To keep it simple, we can just get filters directly as suggested as well.

File: internal/provider/resource_host_set_plugin.go [hunk 2]
```diff
@@ -131,6 +132,25 @@ func setFromHostSetPluginResponseMap(d *schema.ResourceData, raw map[string]inte
 		attrRaw, ok := raw["attributes"]
 		switch ok {
 		case true:
+			if attrMap, ok := attrRaw.(map[string]interface{}); ok {
+				for attrKey := range attrMap {
+					// Flatten attribute filters data structure from Boundary SDK to match terraform input data structure
```
[the remark this answers is not in the document]
Comment (elimt) [reply internal/provider/resource_host_set_plugin.go:137]: I thought about solving it generally by adjusting the data structure of the API/SDK to match what Boundary TF Provider expects and how the CLI even passes values but that could possibly break some existing integrations with our public API.

For restricting the change to just the AWS plugin, the host-catalog has the plugin name/type but I don't think the host-set returns that information. I can crosscheck

**Review by ddebko — APPROVED** [review]
This looks great. Should we add unit tests to validate we are flattening the attribute filters from the Boundary SDK response? Or does the existing unit tests already do this validation?

File: internal/provider/resource_host_set_plugin.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (elimt) [reply internal/provider/resource_host_set_plugin.go:137]: Since the input from TF provider is user provided, I was thinking it could be tricky to change the data structure of that since users can enter almost anything. From the SDK, the data structure we get is consistent and deterministic so it's easier to change.

I can add a comment to explain why we are doing this and what plugin this change applies to so future developers have more context about this change.

I'll add some test coverage for it

File: internal/provider/resource_host_set_plugin.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (louisruch) [reply internal/provider/resource_host_set_plugin.go:150]: Could we not just drop `filters` completely and flatten any attribute that is a `if toFlatten, ok := filtersInt.([]interface{})`

File: internal/provider/resource_host_set_plugin.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (elimt) [reply internal/provider/resource_host_set_plugin.go:137]: I can update it to be very targeted to just filters as you suggested earlier and also add a comment about why we need to flatten this for just aws

I tried to write some tests for the issue but it seems the only accepted `plugin_name` on Boundary TF is `loopback`. `aws` nor `azure` are not accepted. My guess is because of how the controller testing stuff is set up for Boundary. Only allows `loopback` for testing which does not have the issue we are trying to solve for here.
I manually tested it, however.

File: internal/provider/resource_host_set_plugin.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (elimt) [reply internal/provider/resource_host_set_plugin.go:150]: Yes we could definitely do that but I am not sure if we definitely want to flatten all attributes. Just being safer by trying to flatten just filters for now.

**Comment by elimt** [comment]
Closed in favor of https://github.com/hashicorp/boundary/pull/3338. Fixed the issue from Boundary instead.
